// @remotion/studio-server/dist/preview-server/routes/log-update.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.logUpdate = exports.formatPropChange = exports.normalizeQuotes = exports.warnAboutPrettierOnce = void 0;
const renderer_1 = require("@remotion/renderer");
let warnedAboutPrettier = false;
const warnAboutPrettierOnce = (logLevel) => {
    if (warnedAboutPrettier) {
        return;
    }
    warnedAboutPrettier = true;
    renderer_1.RenderInternals.Log.warn({ indent: false, logLevel }, renderer_1.RenderInternals.chalk.yellow('Could not format with Prettier. File will need to be formatted manually.'));
};
exports.warnAboutPrettierOnce = warnAboutPrettierOnce;
const normalizeQuotes = (str) => {
    if (str.length >= 2 &&
        ((str.startsWith("'") && str.endsWith("'")) ||
            (str.startsWith('"') && str.endsWith('"')))) {
        return `'${str.slice(1, -1)}'`;
    }
    return str;
};
exports.normalizeQuotes = normalizeQuotes;
// 24-bit ANSI helpers
const fg = (r, g, b, str) => `\u001b[38;2;${r};${g};${b}m${str}\u001b[39m`;
const bg = (r, g, b, str) => `\u001b[48;2;${r};${g};${b}m${str}\u001b[49m`;
// Monokai-inspired syntax colors
const attrName = (str) => fg(166, 226, 46, str);
const equals = (str) => fg(249, 38, 114, str);
const punctuation = (str) => fg(248, 248, 242, str);
const stringValue = (str) => fg(230, 219, 116, str);
const numberValue = (str) => fg(174, 129, 255, str);
const colorValue = (str) => {
    if ((str.startsWith("'") && str.endsWith("'")) ||
        (str.startsWith('"') && str.endsWith('"'))) {
        return stringValue(str);
    }
    if (/^-?\d+(\.\d+)?$/.test(str)) {
        return numberValue(str);
    }
    return punctuation(str);
};
// Subtle background tints
const removedBg = (str) => bg(80, 20, 20, str);
const addedBg = (str) => bg(30, 80, 30, str);
const colorEnabled = () => renderer_1.RenderInternals.chalk.enabled();
// Format key={value} with Monokai syntax highlighting
const formatSimpleProp = (key, value) => {
    return `${attrName(key)}${equals('=')}${punctuation('{')}${colorValue(value)}${punctuation('}')}`;
};
// Format parentKey={{childKey: value}} with Monokai syntax highlighting
const formatNestedProp = (parentKey, childKey, value) => {
    return `${attrName(parentKey)}${equals('=')}${punctuation('{{')}${punctuation(childKey)}${punctuation(':')} ${colorValue(value)}${punctuation('}}')}`;
};
const formatPropChange = ({ key, oldValueString, newValueString, defaultValueString, }) => {
    if (!colorEnabled()) {
        const dotIdx = key.indexOf('.');
        if (dotIdx === -1) {
            return `${key}={${oldValueString}} \u2192 ${key}={${newValueString}}`;
        }
        const parent = key.slice(0, dotIdx);
        const child = key.slice(dotIdx + 1);
        return `${parent}={{${child}: ${oldValueString}}} \u2192 ${parent}={{${child}: ${newValueString}}}`;
    }
    const isResetToDefault = defaultValueString !== null && newValueString === defaultValueString;
    const isChangeFromDefault = defaultValueString !== null && oldValueString === defaultValueString;
    const dotIndex = key.indexOf('.');
    if (dotIndex === -1) {
        if (isResetToDefault) {
            return removedBg(formatSimpleProp(key, oldValueString));
        }
        if (isChangeFromDefault) {
            return addedBg(formatSimpleProp(key, newValueString));
        }
        return `${removedBg(formatSimpleProp(key, oldValueString))} \u2192 ${addedBg(formatSimpleProp(key, newValueString))}`;
    }
    const parentKey = key.slice(0, dotIndex);
    const childKey = key.slice(dotIndex + 1);
    if (isResetToDefault) {
        return removedBg(formatNestedProp(parentKey, childKey, oldValueString));
    }
    if (isChangeFromDefault) {
        return addedBg(formatNestedProp(parentKey, childKey, newValueString));
    }
    return `${removedBg(formatNestedProp(parentKey, childKey, oldValueString))} \u2192 ${addedBg(formatNestedProp(parentKey, childKey, newValueString))}`;
};
exports.formatPropChange = formatPropChange;
const logUpdate = ({ fileRelativeToRoot, line, key, oldValueString, newValueString, defaultValueString, formatted, logLevel, }) => {
    const locationLabel = `${fileRelativeToRoot}:${line}`;
    const propChange = (0, exports.formatPropChange)({
        key,
        oldValueString: (0, exports.normalizeQuotes)(oldValueString),
        newValueString: (0, exports.normalizeQuotes)(newValueString),
        defaultValueString: defaultValueString !== null ? (0, exports.normalizeQuotes)(defaultValueString) : null,
    });
    renderer_1.RenderInternals.Log.info({ indent: false, logLevel }, `${renderer_1.RenderInternals.chalk.blueBright(`${locationLabel}:`)} ${propChange}`);
    if (!formatted) {
        (0, exports.warnAboutPrettierOnce)(logLevel);
    }
};
exports.logUpdate = logUpdate;
