// @remotion/renderer/dist/options/browser-download-dir.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.browserDownloadDirOption = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const cliFlag = 'browser-download-dir';
let currentDirectory = null;
exports.browserDownloadDirOption = {
    name: 'Browser Download Directory',
    cliFlag,
    description: () => (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: ["Specify the directory where the browser should be downloaded to when calling ",
            jsx_runtime_1.jsx("code", { children: "npx remotion browser ensure" }),
            ". By default, Remotion downloads the browser into ",
            jsx_runtime_1.jsx("code", { children: "node_modules/.remotion" }),
            " inside the nearest project root. Use this option to override the default download location. Pass ",
            jsx_runtime_1.jsx("code", { children: "null" }),
            " to use the default location."] })),
    ssrName: 'browserDownloadDir',
    docLink: 'https://www.remotion.dev/docs/config#setbrowserdownloaddir',
    type: '',
    getValue: ({ commandLine }) => {
        if (commandLine[cliFlag] !== undefined) {
            return {
                source: 'cli',
                value: commandLine[cliFlag],
            };
        }
        if (currentDirectory !== null) {
            return {
                source: 'config',
                value: currentDirectory,
            };
        }
        return {
            source: 'default',
            value: null,
        };
    },
    setConfig: (value) => {
        currentDirectory = value;
    },
    id: cliFlag,
};
